frame/staking/src/slashing.rs

```diff
@@ -315,13 +313,54 @@ fn kick_out_if_recent<T: Config>(params: SlashParams<T>) {
 	if spans.era_span(params.slash_era).map(|s| s.index) == Some(spans.span_index()) {
 		spans.end_span(params.now);
 		<Pallet<T>>::chill_stash(params.stash);
+	}
+
+	// add the validator to the offenders list but since there's no slash being
+	// applied there's no need to disable the validator
+	add_offending_validator::<T>(params.stash, false);
+}
 
-		// make sure to disable validator till the end of this session
-		if T::SessionInterface::disable_validator(params.stash).unwrap_or(false) {
-			// force a new era, to select a new validator set
-			<Pallet<T>>::ensure_new_era()
+/// Add the given validator to the offenders list and optionally disable it.
+/// If after adding the validator `OffendingValidatorsThreshold` is reached
+/// a new era will be forced.
+fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
+	<Pallet<T> as Store>::OffendingValidators::mutate(|offending| {
+		let validator_index =
+			match T::SessionInterface::validators().iter().position(|i| i == stash) {
+				Some(index) => index,
+				None => return,
+			};
+
+		let validator_index_u32 = validator_index as u32;
+
+		match offending.binary_search_by_key(&validator_index_u32, |(index, _)| *index) {
+			// this is a new offending validator
+			Err(index) => {
+				offending.insert(index, (validator_index_u32, disable));
+
+				let offending_threshold = T::OffendingValidatorsThreshold::get() *
+					T::SessionInterface::validators_len() as u32;
+
+				if offending.len() >= offending_threshold as usize {
+					// force a new era, to select a new validator set
+					<Pallet<T>>::ensure_new_era()
```

Comment:
note to follow up here to understand if committing the changes to `offending` is needed in this scenario (as it is doing right now)

Reply:
This might have implications for the election provider. The test case `estimate_next_election_works` should be enough to ensure forcing a new era will not cause any issues.

Reply:
I think it is needed.

forcing a new era doesn't mean that next active session is the first session of a new era.
pallet_session is "delaying" session by one. So the when staking is configured with ForceNewEra then when pallet_session rotate session it will end the session (let's say) 3 (with `end_session(3)`), start the session 4 (with `start_session(4)`) and plan the session 5 (with `new_session(5)`) the session 5 will be the first session of the new era.

So we still need to disable the validators in the `start_session(4)`